enhancements/microshift/microshift-router-configuration.md

```diff
@@ -0,0 +1,515 @@
+means the services are updated by this component. However, ovnk only turns IP
+addresses into iptables rules, it does not take Hostname, which is another
+valid field for the service status.
+
+Since the configuration uses NIC names and also host names, the controller
+needs to translate these to their corresponding IP addresses, and then
+configure them in the service. To keep up with the dynamic nature of IP
+addresses and names, it needs to perform periodic checks and updates.
+
+#### Allowing specific IP addresses
+Allowing specific IP addresses is achieved through the use of `NetworkPolicy`
+resources, as seen in previous examples. These policies include the allowed IP
+addresses and ports in the spec, meaning they reject everything else that is
+not included explicitly.
+
+Due to MicroShift's networking setup, there is one more change required for
+the `NetworkPolicy` to filter out based on the source IP.
+
+As it can be seen in the iptables rules, traffic towards the router ports is
+subject to DNAT to the router service IP and port. All services use
+`externalTrafficPolicy: Cluster`, which will SNAT traffic's source IP to match
+that of the node. This gets further NAT, as described here. By the time the
+`NetworkPolicy` is evaluated the source IP has been replaced with
+`100.64.0.0/16`, which will not produce matches against the configured IP in
+the policy, and traffic will get rejected.
+
+To overcome this, all LoadBalancer services in MicroShift that are subject to
+get NetworkPolicy resources to control external traffic must set
+`externalTrafficPolicy: Local`. This setting will not SNAT packets and the
+source IP is preserved until NetworkPolicy checks run.
+
+This setting, however, comes at the expense of an even load when having more
+than one node, as traffic will not be sent to another node on the same port.
+This is not an issue in single node, which is MicroShift's use case and design
+principle.
+
+This setting is also the default in OCP when deploying in cloud, having a load
+balancer before the cluster. Manifests for both OCP and MicroShift are
+compatible in this regard.
+
+## Open Questions
+N/A
+
+## Test Plan
+All of the changes listed here will be included in the current e2e scenario
+testing harness in MicroShift.
+
+## Graduation Criteria
+Targeting GA for MicroShift 4.16 release.
+
+### Dev Preview -> Tech Preview
+- Ability to utilize the enhancement end to end
+- End user documentation, relative API stability
+- Sufficient test coverage
+
+### Tech Preview -> GA
+- More testing (upgrade, downgrade)
+- Sufficient time for feedback
+- Available by default
+- User facing documentation created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+### Removing a deprecated feature
+N/A
+
+## Upgrade / Downgrade Strategy
+In the previous implementation, using host ports, the router would listen on
+any interface (except loopback). This included some of the internal IP
```

Comment:
Are you sure that MicroShift's router deployment doesn't listen on loopback?  OpenShift's router does, and I'm not sure how you could even configure that without using a custom HAProxy configuration template, which I didn't think MicroShift was doing.

Reply:
If the node host has an eth0 device and an lo device and the router pod uses host networking, I thought the container would share the node host's eth0 and lo.  Does the container in fact get its own lo while sharing the node host's eth0?  

Well, if the statement in the EP is correct, then we don't need to continue this thread (unless you want to); I can go off and investigate how this works for my own edification.

Reply:
I'm still confused as to why the router isn't listening on loopback.  The router binds to all interfaces.  Does the router container not have an lo interface in its network namespace?

Reply:
Will post some examples. All of the info below is from the microshift host itself.
The `hostPort` entries in the router deployment generate this in the host:
```bash
# iptables -t nat -L
...
Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination         
OVN-KUBE-EXTERNALIP  all  --  anywhere             anywhere            
OVN-KUBE-NODEPORT  all  --  anywhere             anywhere            
OVN-KUBE-ITP  all  --  anywhere             anywhere            
KUBE-HOSTPORTS  all  --  anywhere             anywhere             /* kube hostport portals */ ADDRTYPE match dst-type LOCAL
...
Chain KUBE-HOSTPORTS (2 references)
target     prot opt source               destination         
KUBE-HP-M3VSDVN3JKLTMBIB  tcp  --  anywhere             anywhere             /* k8s_router-default-68d4dc8676-n2fz8_openshift-ingress_72872086-65d9-47e6-a787-7b86ecf0d920_0_ hostport 443 */ tcp dpt:https
KUBE-HP-VG7GIYRPKPR5JDKH  tcp  --  anywhere             anywhere             /* k8s_router-default-68d4dc8676-n2fz8_openshift-ingress_72872086-65d9-47e6-a787-7b86ecf0d920_0_ hostport 80 */ tcp dpt:http

Chain KUBE-HP-M3VSDVN3JKLTMBIB (1 references)
target     prot opt source               destination         
KUBE-MARK-MASQ  all  --  10.42.0.7            anywhere             /* k8s_router-default-68d4dc8676-n2fz8_openshift-ingress_72872086-65d9-47e6-a787-7b86ecf0d920_0_ hostport 443 */
DNAT       tcp  --  anywhere             anywhere             /* k8s_router-default-68d4dc8676-n2fz8_openshift-ingress_72872086-65d9-47e6-a787-7b86ecf0d920_0_ hostport 443 */ tcp to:10.42.0.7:443

Chain KUBE-HP-VG7GIYRPKPR5JDKH (1 references)
target     prot opt source               destination         
KUBE-MARK-MASQ  all  --  10.42.0.7            anywhere             /* k8s_router-default-68d4dc8676-n2fz8_openshift-ingress_72872086-65d9-47e6-a787-7b86ecf0d920_0_ hostport 80 */
DNAT       tcp  --  anywhere             anywhere             /* k8s_router-default-68d4dc8676-n2fz8_openshift-ingress_72872086-65d9-47e6-a787-7b86ecf0d920_0_ hostport 80 */ tcp to:10.42.0.7:80
...

# oc get pod -n openshift-ingress -o wide
NAME                              READY   STATUS    RESTARTS   AGE   IP          NODE             NOMINATED NODE   READINESS GATES
router-default-68d4dc8676-n2fz8   1/1     Running   0          87m   10.42.0.7   microshift-dev   <none>           <none>
```
The iptables rules will forward 80 and 443 traffic to the pod IP directly. But there is one more implication with using `hostPort` directives:
```bash
# netstat -ptanu | grep -E ":(80|443)"
tcp        0      0 0.0.0.0:80              0.0.0.0:*               LISTEN      1105/crio           
tcp        0      0 0.0.0.0:443             0.0.0.0:*               LISTEN      1105/crio 
```
Crio is listening in all of the interfaces for any 80/443 traffic too. This means that from the host itself you can reach the router in those ports as well. Trying to connect to any IP in the host works, but there is this nft rule:
```bash
	chain KUBE-FIREWALL {
		ip saddr != 127.0.0.0/8 ip daddr 127.0.0.0/8  ct status dnat counter packets 0 bytes 0 drop
	}
```
Which will drop all attempts to connect to the router from the host.
```bash
# curl -I 192.168.122.254
HTTP/1.0 503 Service Unavailable
pragma: no-cache
cache-control: private, max-age=0, no-cache, no-store
content-type: text/html

[root@microshift-dev microshift]# curl -I 10.43.0.1
^C
[root@microshift-dev microshift]# curl -I 10.44.0.0
HTTP/1.0 503 Service Unavailable
pragma: no-cache
cache-control: private, max-age=0, no-cache, no-store
content-type: text/html

[root@microshift-dev microshift]# curl -I 10.42.0.7
HTTP/1.0 503 Service Unavailable
pragma: no-cache
cache-control: private, max-age=0, no-cache, no-store
content-type: text/html

[root@microshift-dev microshift]# curl -I 127.0.0.1 -m 1
curl: (28) Connection timed out after 1001 milliseconds

[root@microshift-dev microshift]# curl -I 127.0.0.1 -m 5
curl: (28) Connection timed out after 5001 milliseconds
```

From within the router pod everything works:
```bash
# oc exec -it -n openshift-ingress router-default-68d4dc8676-n2fz8 -- curl -I localhost
HTTP/1.0 503 Service Unavailable
pragma: no-cache
cache-control: private, max-age=0, no-cache, no-store
content-type: text/html
```

So maybe I confused you with which localhost I was referring to.

Reply:
Thanks for the detailed explanation!  Maybe it's a question of semantics.  In my mind, the router "listens on loopback" in the sense that it is binding a socket that listens on all interfaces, including lo.  You're saying that any traffic to 127/8 gets dropped by the firewall, which has the same effect as if the router weren't listening on lo, right?  So it doesn't really matter...  Although now I wonder whether there could be surprises in the future if IPv6 were enabled and the firewall blocks 127/8 but not ::1.

Reply:
> You're saying that any traffic to 127/8 gets dropped by the firewall, which has the same effect as if the router weren't listening on lo, right?

Yeah, exazctly. Actually the way it is setup now it does listen on lo, but nft steals the packets before they reach it.

> Although now I wonder whether there could be surprises in the future if IPv6 were enabled and the firewall blocks 127/8 but not ::1.

Right now we do not have support for IPv6, but this is one of the things that we need to look into.